contracts/src/LightClientV2.sol

```diff
@@ -0,0 +1,205 @@
+contract LightClientV2 is LightClient {
+    /// @notice When entering a new epoch and a new stake table snapshot.
+    event NewEpoch(uint64 epoch);
+
+    /// @notice number of blocks per epoch
+    uint64 public _blocksPerEpoch;
+    /// @notice stake table commitments for the current voting stakers
+    StakeTableState public votingStakeTableState;
+
+    /// @notice The last block of the epoch (of the finalizedState) should not be skipped
+    error MissingLastBlockInEpochUpdate();
+    /// @notice Invocation on outdated APIs on V1
+    error DeprecatedApi();
+
+    /// @notice Initialize V2
+    function initializeV2(uint64 blocksPerEpoch) public reinitializer(2) {
+        _initializeV2(blocksPerEpoch);
+    }
+
+    /// @dev Avoid initialization problem during testing, avoid always upgrading from V1
+    function _initializeV2(uint64 blocksPerEpoch) internal {
+        if (blocksPerEpoch == 0) {
+            revert InvalidArgs();
+        }
+        votingStakeTableState = genesisStakeTableState;
+        _blocksPerEpoch = blocksPerEpoch;
+    }
+
+    function getVersion()
+        public
+        pure
+        virtual
+        override
+        returns (uint8 majorVersion, uint8 minorVersion, uint8 patchVersion)
+    {
+        return (2, 0, 0);
+    }
+
+    /// @dev override the V1's to disable calling it
+    function newFinalizedState(LightClientState memory, IPlonkVerifier.PlonkProof memory)
+        external
+        pure
+        override
+    {
+        revert DeprecatedApi();
+    }
+
+    /// @notice identical as LightClientV1's `setstateHistoryRetentionPeriod()`
+    /// but this function name has the correct camelCase
+    function setStateHistoryRetentionPeriod(uint32 historySeconds) public virtual onlyOwner {
+        setstateHistoryRetentionPeriod(historySeconds);
+    }
+
+    /// @dev See detailed doc in `LightClient.sol`
+    /// @param nextStakeTable the stake table to use in the next block (same as the current except
+    function newFinalizedState(
+        LightClientState memory newState,
+        StakeTableState memory nextStakeTable,
+        IPlonkVerifier.PlonkProof memory proof
+    ) external virtual {
+        //revert if we're in permissionedProver mode and the permissioned prover has not been set
+        if (isPermissionedProverEnabled() && msg.sender != permissionedProver) {
+            revert ProverNotPermissioned();
+        }
+
+        if (
+            newState.viewNum <= finalizedState.viewNum
+                || newState.blockHeight <= finalizedState.blockHeight
+        ) {
+            revert OutdatedState();
+        }
+        // format validity check
+        BN254.validateScalarField(newState.blockCommRoot);
+
+        // epoch-related checks
+        uint64 lastUpdateEpoch = currentEpoch();
+        uint64 newEpoch = epochFromBlockNumber(newState.blockHeight, _blocksPerEpoch);
+        if (
+            newEpoch == lastUpdateEpoch + 1 && !isLastBlockInEpoch(finalizedState.blockHeight)
```

Comment:
I think this is a bit cryptic now. I feel like this could be  more readable with separating the case where `lastUpdateEpoch == 0` and then adding some require statements, but not sure.

Reply:
hmm.. i thought about this when I wrote the code. the decision needs to take into consideration of the next `if` as well, if we want `lastUpdateEpoch==0` as a separate case, then most likely we end up with `if else if else else` 3 clauses at least, and I find it harder to read than the current 2 `if` flatten out.